src/librustc_trans/trans/_match.rs

```diff
@@ -233,7 +233,8 @@ struct ConstantExpr<'a>(&'a ast::Expr);
 
 impl<'a> ConstantExpr<'a> {
     fn eq(self, other: ConstantExpr<'a>, tcx: &ty::ctxt) -> bool {
-        match const_eval::compare_lit_exprs(tcx, self.0, other.0, None) {
+        match const_eval::compare_lit_exprs(tcx, self.0, other.0, None,
+                                            |id| {ty::node_id_item_substs(tcx, id).substs}) {
```

Comment:
Well, the other thing is that const_eval has to be working already in `collect`, and I'm trying to get resolution of trait-associated constants working there. We don't have `ClosureTyper` and `ParameterEnvironment` in play at that point, and it would be kind of funky to use them that early.